feature/platform/transceiver/tests/zr_firmware_version_test/zr_firmware_version_test.go

```diff
@@ -0,0 +1,148 @@
+	d := &oc.Root{}
+	i := d.GetOrCreateInterface(dp.Name())
+	i.Enabled = ygot.Bool(enable)
+	i.Type = oc.IETFInterfaces_InterfaceType_ethernetCsmacd
+	gnmi.Replace(t, dut1, gnmi.OC().Interface(dp.Name()).Config(), i)
+	component := opticalChannelComponentFromPort(t, dut1, dp)
+	gnmi.Replace(t, dut1, gnmi.OC().Component(component).OpticalChannel().Config(), &oc.Component_OpticalChannel{
+		TargetOutputPower: ygot.Float64(targetOutputPower),
+		Frequency:         ygot.Uint64(frequency),
+	})
+}
+func verifyFirmwareVersionValue(t *testing.T, dut1 *ondatra.DUTDevice, pStream *samplestream.SampleStream[string]) {
+	firmwareVersionSample := pStream.Next(t)
+	if firmwareVersionSample == nil {
+		t.Fatalf("Firmware telemetry %v was not streamed in the most recent subscription interval", firmwareVersionSample)
+	}
+	firmwareVersionVal, ok := firmwareVersionSample.Val()
+	if !ok {
+		t.Fatalf("Firmware version %q telemetry is not present", firmwareVersionSample)
+	}
+	// Check firmware version return value of correct type
+	if reflect.TypeOf(firmwareVersionVal).Kind() != reflect.String {
+		t.Fatalf("Return value is not type string")
+	}
+}
+
+func TestZRFirmwareVersionState(t *testing.T) {
+	dut1 := ondatra.DUT(t, "dut")
+	dp1 := dut1.Port(t, "port1")
+	dp2 := dut1.Port(t, "port2")
+	t.Logf("dut1: %v", dut1)
+	t.Logf("dut1 dp1 name: %v", dp1.Name())
+	configInterface(t, dut1, dp1, true)
+	configInterface(t, dut1, dp2, true)
+	gnmi.Await(t, dut1, gnmi.OC().Interface(dp1.Name()).OperStatus().State(), time.Minute, oc.Interface_OperStatus_UP)
+	transceiverName := gnmi.Get(t, dut1, gnmi.OC().Interface(dp1.Name()).Transceiver().State())
+	// Check if TRANSCEIVER is of type 400ZR
+	if dp1.PMD() != ondatra.PMD400GBASEZR {
+		t.Fatalf("%s Transceiver is not 400ZR its of type: %v", transceiverName, dp1.PMD())
+	}
+	component1 := gnmi.OC().Component(transceiverName)
+
+	p1Stream := samplestream.New(t, dut1, component1.FirmwareVersion().State(), 10*time.Second)
+
+	verifyFirmwareVersionValue(t, dut1, p1Stream)
+
+	p1Stream.Close()
+}
+func TestZRFirmwareVersionStateInterfaceFlap(t *testing.T) {
+	dut1 := ondatra.DUT(t, "dut")
+	dp1 := dut1.Port(t, "port1")
+	dp2 := dut1.Port(t, "port2")
+	t.Logf("dut1: %v", dut1)
+	t.Logf("dut1 dp1 name: %v", dp1.Name())
+	configInterface(t, dut1, dp1, true)
+	configInterface(t, dut1, dp2, true)
+	gnmi.Await(t, dut1, gnmi.OC().Interface(dp1.Name()).OperStatus().State(), time.Minute, oc.Interface_OperStatus_UP)
+	transceiverName := gnmi.Get(t, dut1, gnmi.OC().Interface(dp1.Name()).Transceiver().State())
+	// Check if TRANSCEIVER is of type 400ZR
+	if dp1.PMD() != ondatra.PMD400GBASEZR {
+		t.Fatalf("%s Transceiver is not 400ZR its of type: %v", transceiverName, dp1.PMD())
+	}
+	// Disable interface
+	configInterface(t, dut1, dp1, false)
+	component1 := gnmi.OC().Component(transceiverName)
+
+	p1Stream := samplestream.New(t, dut1, component1.FirmwareVersion().State(), 10*time.Second)
```

Comment:
Minor optimization here, but did you want to start the sample stream before disabling the interface? This was you capture all transitions. (UP->DOWN and DOWN->UP).  As written, the streaming for firmware begins after the interface is configured DOWN.

Instead maybe the flow should be:

1. Configure DUT interface, Await OperStatus_UP
2. Start Stream 
3. Verify firmware version from stream
4. Set DUT interface down, Await OperStatus_DOWN
5. Verify firmware version from stream
6. Set DUT interface up, Await OperStatus_UP
7. Verify firmware version from stream